### PR TITLE
Fetch adjectives for nouns after filtering by IDF

### PR DESCRIPTION
**Pull request by guptashashij**
Fixes #<issue_number_goes_here>

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

File: adj_idf_gen.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+"""Generate IDF for adjectives"""
+import pickle
+from nltk.corpus import wordnet
+from nltk.text import TextCollection, Text
+f = open("generated_files/all.txt", "r")
```
Comment (sharsh-google) [inline adj_idf_gen.py]: leave a blank line after import

File: README.md [hunk 2]
```diff
@@ -9,6 +9,13 @@ adjectives for noun. (eg. [good adjectves for babies], [adjectives for tatoos]
 etc.). We would like to build a system that can categorize the adjectives for a
 noun as [positive], [negative] and by property [size], [beauty], [speed] etc.
 
+
```
Comment (sharsh-google) [inline README.md:12]: Move this information to a new file setup.md We may need to add some more setup info to it later.

File: adj_idf_gen.py (hunk 1 above)
Comment (sharsh-google) [inline adj_idf_gen.py]: give a better name to the file. "all.txt" is ambiguous? What "all" does it contain? Use something like books_data.txt

File: adj_idf_gen.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+lines = f.readlines()
+tokens = []
+for line in lines:
```
Comment (sharsh-google) [inline adj_idf_gen.py]: separate code blocks by blank lines. Here and everywhere. LIke blank line after import, blanks line before loop, after loop. Group code which does a simple job together and separate each such blocks.

File: adj_idf_gen.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    line = line.lower()
+    line = line.split("\n")[0].split(" ")
+    tokens.extend(line)
+text = Text(tokens)
+mytexts = TextCollection(text)
```
Comment (sharsh-google) [inline adj_idf_gen.py]: text, mytext are not good variable names. Use specific names which represent what is contained in those variables.

File: adj_idf_gen.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+adj_idf = {}
+for synset in list(wordnet.all_synsets(wordnet.ADJ)):
```
Comment (sharsh-google) [inline adj_idf_gen.py]: Avoid using api names as variable names. for adjective_set..?

File: bert_setup.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        bert_tokenizer : Tokenize a sentence using Bert_Tokenizer.
+        bert_model : Predict new words for a token in the sentence.
+    """
+    def __init__(self):
+        self.tokenizer = TreebankWordTokenizer()
+        self.bert_tokenizer = Bert_Tokenizer.from_pretrained('bert-large-uncased')
+        self.bert_model = BertForMaskedLM.from_pretrained('bert-large-uncased')
+        self.bert_model.eval()
+
+    def perturb_bert(self, sentence, num, masked_index):
```
Comment (sharsh-google) [inline bert_setup.py]: nit: num_suggest

File: main.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+"""Generate csv file with nouns as first coloumn and other columns correspond to adjectives"""
```
Comment (sharsh-google) [inline main.py]: Rename the file to something like noun_to_adj_main.py

File: bert_setup.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        masked_index : the index of the word in the sentence for which new
+        words need to be suggested
+
+        Returns:
+        A dictionary with new words suggested as keys and their confidence scores
+        as values.
+        """
+        tokens = self.tokenizer.tokenize(sentence)
+
+        # for each idx, use Bert to generate k (i.e., num) candidate tokens
```
Comment (sharsh-google) [inline bert_setup.py]: What is the casing convention for comments? First letter capital? Whatever it is, please follow and be consistent throughout the file.

File: bert_setup.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        low_tokens[masked_index] = '[MASK]'
+        # Eliminating cases for "'s" as Bert does not work well on these cases.
+        if original_word == "'s":
+            return {}
+        # try whether all the tokens are in the vocabulary
+        try:
+            indexed_tokens = self.bert_tokenizer.convert_tokens_to_ids(low_tokens)
+            prediction = self.bert_model(torch.tensor([indexed_tokens]))
+
+        except KeyError:
```
Comment (sharsh-google) [inline bert_setup.py]: In production, I'd ideally log the error here. Can you at least add a comment about the error here?

File: bert_setup.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            return {}
+
+        # get the similar words
+        topk_idx = torch.topk(prediction[0, masked_index], num)[1].tolist()
+        scores = torch.topk(prediction[0, masked_index], num)[0].tolist()
+        topk_tokens = self.bert_tokenizer.convert_ids_to_tokens(topk_idx)
+        # generate similar sentences
+        tokens[masked_index] = original_word
+        token_score = {}
```
Comment (sharsh-google) [inline bert_setup.py]: use better names: token_score is not suggestive of the return data. Same goes for other variable names here.

File: utils.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
+"""Some basic functions required by COntextual Adjectives"""
```
Comment (sharsh-google) [inline utils.py]: Contenxtual casing

File: utils.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+from nltk.corpus import wordnet
+
+def sentence_list_gen(filename):
+    """read lines from a file and make a list of lines.
```
Comment (sharsh-google) [inline utils.py]: According to style guide the function comments should be descriptive and not imperative. Like "Reads lines from a file" and not "Read lines". Please update here and everywhere.

File: utils.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    Args:
+      noun_to_adj : Given dictionary of nouns to adjectives with BERT score.
+      adj_idf : IDF values of adjectives.
+      threshold : Threshold to filter by IDF.
+      max_adj : Max number of adjectives for a noun.
+
+    Returns:
+      A Dictionary of noun to adj after filtering in sorted order of BERT score.
+    """
+    new_dic = {}
```
Comment (sharsh-google) [inline utils.py]: Do not use variable names like new_dic, temp etc. Use names which are specific and describe the data contained in them.

File: nohup.out [hunk 14]
```diff
@@ -0,0 +1,30 @@
+2020-06-02 07:50:38.257077: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
```
Comment (sharsh-google) [inline nohup.out]: What is this file? What does it contain? Do we need to check it in?

File: utils.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+                sorted_lis.append((key, -1*val))
+        new_dic[noun] = sorted_lis
+    return new_dic
+
+def save_as_csv(filename, noun_to_adj):
+    """Save a dictionary noun_to_adj to a given filename as csv."""
+    file = open(filename, "w")
+    for noun in noun_to_adj.keys():
+        if noun_to_adj[noun] != []:
+            sent = noun + "\t"
```
Comment (sharsh-google) [inline utils.py]: do you mean sentence by sent? use descriptive names.

File: bert_setup.py (hunk 10 above)
Comment (guptashashij) [reply bert_setup.py]: I think it is suggestive if return data. It is a dictionary of token and their BERT score.

File: utils.py (hunk 12 above)
Comment (guptashashij) [reply utils.py]: I didn't get this.

Comment (sharsh-google) [reply utils.py]: I mean the function comments should be of descriptive form like "Reads the data..", "Returns the value.." and not imperative "Read the data", "Return the value". 
Read this for details http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: noun_to_adj_main.py [hunk 16]
```diff
@@ -25,4 +26,9 @@
 # filtering by IDF (threshold = 8.125) , sorting by BERT score and returning top k=20 adjectives.
```
Comment (sharsh-google) [inline noun_to_adj_main.py]: correct the casing: Filtering

File: noun_to_adj_main.py [hunk 17]
```diff
@@ -13,11 +13,11 @@
 generator = NounToAdjGen(noun_list, adj_list) # an instance of class noun_to_adj_gen
 
 # 10 = Number of perturbations you want to make for a word in a sentence
-generator.add_to_dic(sentences[:100], 10) 
+generator.add_to_dic(sentences, 10) 
 
-# Saving the dictionary to a pickle file
-# with open("generated_files/noun_to_adj_score.dat", 'wb') as f:
-#     pickle.dump(getattr(generator, 'noun_to_adj'), f)
+#Saving the dictionary to a pickle file
```
Comment (sharsh-google) [inline noun_to_adj_main.py]: Space after # and comment. Here and everywhere.

**Review by sharsh-google — APPROVED** [review]

File: bert_setup.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        # Error when invalid index tokens.    
+        except KeyError:
+            return {}
+
+        # Get the similar words
+        topk_idx = torch.topk(prediction[0, masked_index], num_suggest)[1].tolist()
+        scores = torch.topk(prediction[0, masked_index], num_suggest)[0].tolist()
+        topk_tokens = self.bert_tokenizer.convert_ids_to_tokens(topk_idx)
+        # Generate similar sentences
+        tokens[masked_index] = original_word
```
Comment (sowjanyavemuri-google) [inline bert_setup.py]: why do we need this line?

File: bert_setup.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        # For each idx, use Bert to generate k (i.e., num_suggest) candidate tokens
+        original_word = tokens[masked_index]
+        # Getting the base form of the word to check for it's synonyms
+        low_tokens = [x.lower() for x in tokens]
+        low_tokens[masked_index] = '[MASK]'
+        # Eliminating cases for "'s" as Bert does not work well on these cases.
+        if original_word == "'s":
+            return {}
+        # Try whether all the tokens are in the vocabulary
+        try:
```
Comment (sowjanyavemuri-google) [inline bert_setup.py]: Does Bert have an OOV token i.e assigning an oov token for all non-vocabulary words. In that case this is not needed right?

File: noun_to_adj_gen.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    def __init__(self, noun_list, adj_list):
+        """Initializing noun to adjective dictionary."""
+        self.noun_to_adj = {}
+        for noun in noun_list:
+            self.noun_to_adj[noun] = []
+        # Use nltk treebank tokenizer
+        self.tokenizer = TreebankWordTokenizer()
+        # Initializing the bert class
+        self.bert_model = Bert()
+        self.adj_tags = ['JJ', 'JJR', 'JJS']
```
Comment (sowjanyavemuri-google) [inline noun_to_adj_gen.py]: Can you add a comment mentioning the source of these tags (probably add a link to wordnet pos tags page)?

File: noun_to_adj_gen.py [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+        self.noun_tags = ['NN', 'NNS', 'NNP', 'NNPS']
+        self.noun_list = noun_list
+        self.adj_list = adj_list
+
+    def add_to_dic(self, sentences, num_of_perturb):
```
Comment (sowjanyavemuri-google) [inline noun_to_adj_gen.py]: nit: add_to_dictionary*

File: bert_setup.py (hunk 18 above)
Comment (guptashashij) [reply bert_setup.py]: Which line?

File: bert_setup.py (hunk 19 above)
Comment (guptashashij) [reply bert_setup.py]: I don't think it has. BERT gives error when passed words like 's

File: bert_setup.py (hunk 18 above)
Comment (sowjanyavemuri-google) [reply bert_setup.py]: line 56.
I don't see tokens object used anywhere post this.

**Review by sowjanyavemuri-google — APPROVED** [review]

**Pull request merged**